src/appleseed/renderer/modeling/frame/frame.cpp

```diff
@@ -462,28 +490,465 @@ void Frame::denoise(
+    if (!is_checkpoint_compatible(impl->m_checkpoint_resume_path, *this, checkpoint_props))
+    {
+        reader.close();
+        return false;
+    }
+
+    const size_t start_pass = get<2>(checkpoint_props[0]).get<size_t>("appleseed:LastPass") + 1;
+
+    // Check if passes have already been rendered.
+    if (impl->m_pass_count <= start_pass)
+    {
+        RENDERER_LOG_WARNING("passes have previously been rendered and saved into the checkpoint");
+        reader.close();
+        return false;
+    }
+
+    // Interface the shading buffer in a canvas.
+    ShadingBufferCanvas shading_canvas(*this, buffer_factory);
+
+    // Load tiles from the checkpoint.
+    const CanvasProperties& beauty_props = get<1>(checkpoint_props[0]);
+
+    for (size_t tile_y = 0; tile_y < beauty_props.m_tile_count_y; ++tile_y)
+    {
+        for (size_t tile_x = 0; tile_x < beauty_props.m_tile_count_x; ++tile_x)
+        {
+            // Read shading buffer.
+            reader.choose_subimage(1);
+            Tile& shading_tile = shading_canvas.tile(tile_x, tile_y);
+            reader.read_tile(tile_x, tile_y, &shading_tile);
+
+            // No need to read beauty and filtered AOVs because they
+            // are in the shading buffer.
+
+            // Read Unfiltered AOV layers.
+            vector<unique_ptr<Tile>> aov_tiles;
+            vector<const float*> aov_ptrs;
+            for (size_t i = 0; i < aovs().size(); ++i)
+            {
+                UnfilteredAOV* aov = dynamic_cast<UnfilteredAOV*>(
+                    aovs().get_by_index(i));
+
+                if (aov == nullptr)
+                    continue;
+
+                const string aov_name = aov->get_name();
+
+                // Search layer index in the file.
+                size_t subimage_index(~0);
+                for (size_t s = 0; s < checkpoint_props.size(); ++s)
+                {
+                    if (get<0>(checkpoint_props[s]) == aov_name)
+                    {
+                        subimage_index = s;
+                        break;
+                    }
+                }
+
+                assert(subimage_index != size_t(~0));
+
+                Image& aov_image = aov->get_image();
+                Tile& aov_tile = aov_image.tile(tile_x, tile_y);
+                reader.choose_subimage(subimage_index);
+                reader.read_tile(tile_x, tile_y, &aov_tile);
+            }
+        }
+    }
+
+    reader.close();
+
+    // Load internal AOVs (from external files).
+    for (size_t i = 0, e = internal_aovs().size(); i < e; ++i)
+    {
+        AOV* aov = internal_aovs().get_by_index(i);
+
+        DenoiserAOV* denoiser_aov = dynamic_cast<DenoiserAOV*>(aov);
+
+        // Save denoiser checkpoint.
+        if (denoiser_aov != nullptr)
+        {
+            if (!load_denoiser_checkpoint(impl->m_checkpoint_resume_path, denoiser_aov))
+                return false;
+        }
+    }
+
+    RENDERER_LOG_INFO("opened checkpoint resuming at pass %s",
```

Comment:
Statement fits on a single line.

Reply:
End message with a period.

Let's mention the checkpoint file path:

    read checkpoint file %s, resuming rendering at pass %s.